--- gtk+/gtk/gtktreeview.c	(revision 1023)
+++ gtk+/gtk/gtktreeview.c	(revision 1024)
@@ -9364,6 +9364,8 @@
            !_gtk_tree_selection_is_row_selectable (tree_view->priv->selection,
                                                    new_cursor_node,
                                                    new_cursor_path));
+  if (new_cursor_path)
+    gtk_tree_path_free (new_cursor_path);
 
   /*
    * If the list has only one item and multi-selection is set then select
