--- gtk+/gtk/gtktreemodelfilter.c	(revision 1239)
+++ gtk+/gtk/gtktreemodelfilter.c	(revision 1240)
@@ -2493,7 +2493,10 @@
       if (!filter->priv->root)
         gtk_tree_model_filter_build_level (filter, NULL, NULL, FALSE);
 
-      return FILTER_LEVEL (filter->priv->root)->visible_nodes;
+      if (filter->priv->root)
+        return FILTER_LEVEL (filter->priv->root)->visible_nodes;
+
+      return 0;
     }
 
   elt = FILTER_ELT (iter->user_data2);
