diff -ur glib-2.6.2/ChangeLog glib-2.6.2.new/ChangeLog
--- glib-2.6.2/ChangeLog	2005-02-04 07:09:47.000000000 +0200
+++ glib-2.6.2.new/ChangeLog	2005-08-08 16:03:41.772396936 +0300
@@ -1,3 +1,14 @@
+2005-08-05  Tommi Komulainen  <tommi.komulainen@nokia.com>
+	* glib/gscanner.h(struct _GScanner): Add cache_end for quick access.
+
+	* glib/gscanner.c(g_scanner_get_token_cache): Avoid addressing bytes
+        beyond the end of the mmap'd area.
+        Fixes #16829 GScanner cache file access out of bounds
+
+	* glib/gscanner.c(osso_g_scanner_cache_open): If open fails, do not
+        try to close the invalid file descriptor.
+        Fixes #16264 theme-engine closes invalid (already closed?) file descriptors
+
 2005-02-02  Matthias Clasen  <mclasen@redhat.com>
 
 	* === Released 2.6.2 ===
