--- dbus/dbus/dbus-pending-call.h	2005-02-18 20:02:28.000000000 +0200
+++ dbus.work/dbus/dbus-pending-call.h	2005-05-13 14:28:17.578528048 +0300
@@ -41,6 +41,8 @@
                                               DBusFreeFunction               free_user_data);
 void         dbus_pending_call_cancel        (DBusPendingCall               *pending);
 dbus_bool_t  dbus_pending_call_get_completed (DBusPendingCall               *pending);
+DBusMessage* dbus_pending_call_get_reply (DBusPendingCall *pending);
+
 DBusMessage* dbus_pending_call_steal_reply   (DBusPendingCall               *pending);
 void         dbus_pending_call_block         (DBusPendingCall               *pending);
 
