diff -ur dbus/bus/desktop-file.c dbus.new/bus/desktop-file.c
--- dbus/bus/desktop-file.c	2006-01-11 16:33:23.000000000 +0200
+++ dbus.new/bus/desktop-file.c	2006-01-30 18:21:22.902824816 +0200
@@ -466,6 +466,9 @@
   while (p < line_end &&
 	 (valid[_dbus_string_get_byte (&parser->data, p)] & VALID_KEY_CHAR))
     p++;
+  /* skip possible white space in end of the key */
+  while (isspace (_dbus_string_get_byte (&parser->data, p)))
+    --p;
   key_end = p;
   
   if (key_start == key_end)
