diff -ur dbus/bus/system.conf.in dbus.new/bus/system.conf.in
--- dbus/bus/system.conf.in	2005-11-04 14:04:20.000000000 +0200
+++ dbus.new/bus/system.conf.in	2005-11-07 17:23:20.656612904 +0200
@@ -20,6 +20,8 @@
   <!-- Fork into daemon mode -->
   <fork/>
 
+  <servicedir>@EXPANDED_DATADIR@/dbus-1/services</servicedir>
+
   <!-- Write a pid file -->
   <pidfile>@DBUS_SYSTEM_PID_FILE@</pidfile>
 
Only in dbus.new/bus: system.conf.in~
Only in dbus/debian/patches: config.diff
