diff -pur dbus-1.0.0/dbus/dbus-sysdeps.h dbus-1.0.0.new/dbus/dbus-sysdeps.h
--- dbus-1.0.0/dbus/dbus-sysdeps.h	2006-11-02 06:26:23.000000000 +0200
+++ dbus-1.0.0.new/dbus/dbus-sysdeps.h	2006-11-13 17:25:21.118295816 +0200
@@ -67,8 +67,12 @@ typedef struct DBusList DBusList;
 #if     __GNUC__ > 2 || (__GNUC__ == 2 && __GNUC_MINOR__ > 4)
 #define _DBUS_GNUC_PRINTF( format_idx, arg_idx )    \
   __attribute__((__format__ (__printf__, format_idx, arg_idx)))
+#ifndef __arm__
 #define _DBUS_GNUC_NORETURN                         \
   __attribute__((__noreturn__))
+#else  /* !__arm__ */
+#define _DBUS_GNUC_NORETURN
+#endif
 #else   /* !__GNUC__ */
 #define _DBUS_GNUC_PRINTF( format_idx, arg_idx )
 #define _DBUS_GNUC_NORETURN
