diff -ur dbus/dbus/dbus-auth.c dbus.new/dbus/dbus-auth.c
--- dbus/dbus/dbus-auth.c	2005-11-04 14:04:14.000000000 +0200
+++ dbus.new/dbus/dbus-auth.c	2005-11-09 16:05:23.823091256 +0200
@@ -1059,11 +1059,11 @@
       _dbus_verbose ("%s: credentials uid="DBUS_UID_FORMAT
                      " gid="DBUS_GID_FORMAT
                      " do not allow uid="DBUS_UID_FORMAT
-                     " gid="DBUS_GID_FORMAT"\n",
+                     " gid="DBUS_GID_FORMAT", but I'm hacked to ignore that\n",
                      DBUS_AUTH_NAME (auth),
                      auth->credentials.uid, auth->credentials.gid,
                      auth->desired_identity.uid, auth->desired_identity.gid);
-      return send_rejected (auth);
+      return TRUE;
     }
 }
 
Only in dbus.new/dbus: dbus-auth.c~
Only in dbus/debian/patches: evil-security-hack.diff
