Index: glib/dbus-gproxy.c
===================================================================
--- glib/dbus-gproxy.c	(.../ags/dbus/0.61-osso2)	(revision 2887)
+++ glib/dbus-gproxy.c	(.../runk/dbus)	(revision 2887)
@@ -493,6 +493,8 @@
   const DBusGProxyNameOwnerInfo *info = a;
   const char *name = b;
 
+  if (info == NULL || info->name == NULL)
+    return 1;
   return strcmp (info->name, name);
 }
 
