diff -ur ../../dbus/dbus/dbus-bus.c dbus/dbus-bus.c
--- ../../dbus/dbus/dbus-bus.c	2006-03-03 19:26:22.000000000 +0200
+++ dbus/dbus-bus.c	2006-03-23 18:26:36.660518120 +0200
@@ -29,6 +29,9 @@
 #include "dbus-marshal-validate.h"
 #include "dbus-threads-internal.h"
 #include <string.h>
+#include <glib-2.0/glib.h>
+#include <glib-2.0/glib-object.h>
+#include <dbus/dbus-glib.h>
 
 /**
  * @defgroup DBusBus Message bus APIs
@@ -314,6 +317,11 @@
   _dbus_return_val_if_fail (type >= 0 && type < N_BUS_TYPES, NULL);
   _dbus_return_val_if_error_is_set (error, NULL);
 
+  /* evil init threads hack */
+  if (!g_thread_supported())
+    g_thread_init(NULL);
+  dbus_g_thread_init();
+
   _DBUS_LOCK (bus);
 
   if (!init_connections_unlocked ())
diff -ur ../../dbus/glib/dbus-gmain.c glib/dbus-gmain.c
--- ../../dbus/glib/dbus-gmain.c	2006-03-03 19:26:22.000000000 +0200
+++ glib/dbus-gmain.c	2006-03-23 18:26:19.935060776 +0200
@@ -567,6 +567,11 @@
 {
   ConnectionSetup *old_setup;
   ConnectionSetup *cs;
+
+  /* evil init threads hack */
+  if (!g_thread_supported())
+    g_thread_init(NULL);
+  dbus_g_thread_init();
   
   /* FIXME we never free the slot, so its refcount just keeps growing,
    * which is kind of broken.
@@ -754,7 +759,7 @@
   g_return_val_if_fail (error == NULL || *error == NULL, NULL);
 
   _dbus_g_value_types_init ();
-  
+
   dbus_error_init (&derror);
 
   connection = dbus_bus_get (type, &derror);
