Index: glib/dbus-gobject.c
===================================================================
RCS file: /cvs/dbus/dbus/glib/dbus-gobject.c,v
retrieving revision 1.53
diff -u -B -p -r1.53 dbus-gobject.c
--- glib/dbus-gobject.c	27 Nov 2005 16:55:09 -0000	1.53
+++ glib/dbus-gobject.c	9 May 2006 11:56:26 -0000
@@ -1732,6 +1732,13 @@ funcsig_equal (gconstpointer aval,
   return TRUE;
 }
 
+static void
+funcsig_free (DBusGFuncSignature *sig)
+{
+  g_free (sig->params);
+  g_free (sig);
+}
+
 GClosureMarshal
 _dbus_gobject_lookup_marshaller (GType        rettype,
 				 guint        n_params,
@@ -1865,7 +1872,7 @@ dbus_g_object_register_marshaller_array 
   if (marshal_table == NULL)
     marshal_table = g_hash_table_new_full (funcsig_hash,
 					   funcsig_equal,
-					   g_free,
+					   (GDestroyNotify) funcsig_free,
 					   NULL);
   sig = g_new0 (DBusGFuncSignature, 1);
   sig->rettype = G_TYPE_FUNDAMENTAL (rettype);
