--- dbus/glib/dbus-binding-tool-glib.c-orig	2006-02-09 00:31:45.000000000 +0100
+++ dbus/glib/dbus-binding-tool-glib.c	2006-02-09 00:33:04.000000000 +0100
@@ -1072,7 +1072,7 @@
           g_string_append_printf (string, ", %s", value_lookup);
           g_free (value_lookup);
         }
-      g_string_append (string, ")");
+      g_string_append (string, ", G_TYPE_INVALID)");
       return g_string_free (string, FALSE);
     }
  
