Index: dbus/glib/dbus-gvalue.c
===================================================================
--- dbus.orig/glib/dbus-gvalue.c	2006-02-07 08:50:59.000000000 +0000
+++ dbus/glib/dbus-gvalue.c	2006-02-07 09:19:57.000000000 +0000
@@ -1081,9 +1081,10 @@
   dbus_message_iter_get_fixed_array (&subiter,
 				     &msgarray,
 				     &msgarray_len);
-  g_assert (msgarray != NULL);
-  g_assert (msgarray_len >= 0);
-  g_array_append_vals (ret, msgarray, (guint) msgarray_len);
+  g_assert (msgarray != NULL || msgarray_len == 0);
+
+  if (msgarray_len)
+    g_array_append_vals (ret, msgarray, (guint) msgarray_len);
 
   g_value_set_boxed_take_ownership (value, ret);
   
