diff -u -d -r1.5 -r1.6
--- glib/dbus-gtype-specialized.c	13 Feb 2006 22:30:11 -0000	1.5
+++ glib/dbus-gtype-specialized.c	7 Jun 2006 00:31:01 -0000	1.6
@@ -392,7 +392,7 @@
                         ...)
 {
   GArray *types;
-  GType curtype;
+  GType curtype, ret;
   va_list args;
   va_start (args, first_type);
 
@@ -404,8 +404,13 @@
       curtype = va_arg (args, GType);
     }
   va_end (args);
-  return lookup_or_register_specialized (container, types->len, (GType*)types->data);
 
+  ret = lookup_or_register_specialized (container, types->len,
+      (GType *) types->data);
+
+  g_array_free (types, TRUE);
+
+  return ret;
 }
