diff -ur 0.23.4-8osso/bus/config-parser.c 0.23.4-8osso.new/bus/config-parser.c
--- 0.23.4-8osso/bus/config-parser.c	2005-09-05 13:14:53.278274000 +0300
+++ 0.23.4-8osso.new/bus/config-parser.c	2005-11-28 14:42:32.570036240 +0200
@@ -399,7 +399,7 @@
        * message if something screws up, but making it too short means
        * they might see a false failure.
        */
-      parser->limits.activation_timeout = 25000; /* 25 seconds */
+      parser->limits.activation_timeout = 32000; /* 32 seconds */
 
       /* Making this long risks making a DOS attack easier, but too short
        * and legitimate auth will fail.  If interactive auth (ask user for
