diff -ur dbus/tools/dbus-print-message.c dbus.work/tools/dbus-print-message.c
--- dbus/tools/dbus-print-message.c	2005-02-11 01:31:29.000000000 +0200
+++ dbus.work/tools/dbus-print-message.c	2005-06-10 14:32:47.678450696 +0300
@@ -53,11 +53,13 @@
     {
     case DBUS_MESSAGE_TYPE_METHOD_CALL:
     case DBUS_MESSAGE_TYPE_SIGNAL:
-      printf ("%s interface=%s; member=%s; sender=%s\n",
+      printf ("%s if=%s; mtd=%s; snd=%s; dtl=%s\n",
               type_to_name (message_type),
               dbus_message_get_interface (message),
               dbus_message_get_member (message),
-              sender ? sender : "(no sender)");
+              sender ? sender : "(no sender)",
+	      dbus_message_has_detail(message) ?
+	      dbus_message_get_detail(message) : "n/a");
       break;
       
     case DBUS_MESSAGE_TYPE_METHOD_RETURN:
Only in dbus.work/tools/: dbus-print-message.c~
