diff -ur ../dbus/dbus/dbus-string.c dbus/dbus/dbus-string.c
--- ../dbus/dbus/dbus-string.c	2005-09-05 13:11:12.000000000 +0300
+++ dbus/dbus/dbus-string.c	2005-12-15 17:31:10.356857816 +0200
@@ -423,6 +423,24 @@
 {
   /* Note, we are setting the length not including nul termination */
 
+  /* Kimmo's MEMORY-SAVING HACK: */
+  if (real->allocated > new_length * 2 + ALLOCATION_PADDING)
+    {
+       /* free some space */
+       int new_size = new_length * 2 + ALLOCATION_PADDING;
+       char* new_str = dbus_realloc (real->str - real->align_offset,
+                                     new_size);
+       if (_DBUS_UNLIKELY (new_str == NULL))
+         return FALSE;
+
+       real->str = new_str + real->align_offset;
+       real->allocated = new_size;
+       fixup_alignment (real);
+       real->len = new_length;
+       real->str[new_length] = '\0';
+       return TRUE;
+    }
+
   /* exceeding max length is the same as failure to allocate memory */
   if (_DBUS_UNLIKELY (new_length > real->max_length))
     return FALSE;
Only in dbus/dbus: dbus-string.c~
