diff -Naur fontconfig/fonts.conf.in fontconfig-2.3.95-optimize-config/fonts.conf.in
--- fontconfig/fonts.conf.in	2006-04-15 03:24:26.000000000 +0300
+++ fontconfig-2.3.95-optimize-config/fonts.conf.in	2006-07-12 15:15:15.000000000 +0300
@@ -72,6 +72,7 @@
   Serif faces
  -->
 	<alias>
+		<family>NewTimes</family>
 		<family>DejaVu Serif</family>
 		<family>Bitstream Vera Serif</family>
 		<family>Times New Roman</family>
@@ -92,6 +93,8 @@
   Sans-serif faces
  -->
 	<alias>
+		<family>SwissA</family>
+		<family>Nokia Sans</family>
 		<family>DejaVu Sans</family>
 		<family>Bitstream Vera Sans</family>
 		<family>Helvetica</family>
@@ -114,6 +117,7 @@
   Monospace faces
  -->
  	<alias>
+		<family>NewCourier</family>
 		<family>DejaVu Sans Mono</family>
 		<family>Courier</family>
 		<family>Courier New</family>
@@ -146,6 +150,7 @@
 <!--
   URW provides metric and shape compatible fonts for these 10 Adobe families.
   -->
+<!-- Removing as irrelevant for stock device
         <alias>
           <family>Avant Garde</family>
           <accept><family>URW Gothic L</family></accept>
@@ -190,10 +195,13 @@
 	    <string>Standard Symbols L</string>
 	  </edit>
 	</match>
+-->	
+	
 <!--
  AMT provides metric and shape compatible fonts for these three web font
  families.
  -->
+ <!-- Removing as irrelevant for stock device
  	<alias>
 		<family>Times New Roman</family>
 		<accept><family>Thorndale AMT</family></accept>
@@ -206,12 +214,136 @@
 		<family>Courier New</family>
 		<accept><family>Cumberland AMT</family></accept>
 	</alias>
+-->
+
+<!--
++  Provide aliases for the content fonts
++-->
+	<alias>
+		<family>Arial</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>ArialNarrow</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>ArialRounded</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Caflish Script Web</family>
+		<accept><family>NewTimes</family></accept> <!-- NewTimes Italic -->
+	</alias>
+	<alias>
+		<family>Chicago</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Comic Sans MS</family>
+		<accept><family>SwissA</family></accept> <!-- SwissA Italic -->
+	</alias>
+	<alias>
+		<family>Courier</family>
+		<accept><family>NewCourier</family></accept>
+	</alias>
+	<alias>
+		<family>Geneva</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Georgia</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias>
+		<family>Giddyup Web</family>
+		<accept><family>SwissA</family></accept> <!-- SwissA Italic -->
+	</alias>
+	<alias>
+		<family>Helvetica</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>HelveticaNeue</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Impact</family>
+		<accept><family>SwissA</family></accept> <!-- SwissA Bold  -->
+	</alias>
+	<alias>
+		<family>Mezz Web</family>
+		<accept><family>SwissA</family></accept> <!-- SwissA Italic -->
+	</alias>
+	<alias>
+		<family>Minion Web</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias>
+		<family>Monaco</family>
+		<accept><family>NewCourier</family></accept>
+	</alias>
+	<alias>
+		<family>Myriad Web</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Myriad Condenced Web</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>New York</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias>
+		<family>Palatino</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias>
+		<family>Penumbra Web</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Symbol</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias>
+		<family>Tahoma</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Times New Roman</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias> <!-- Additional one to cover plain "Times" -->
+		<family>Times</family>
+		<accept><family>NewTimes</family></accept>
+	</alias>
+	<alias>
+		<family>Trebuchet MS</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Verdana</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Webdings</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+	<alias>
+		<family>Windings</family>
+		<accept><family>SwissA</family></accept>
+	</alias>
+
  
 <!--
   Some Asian fonts misadvertise themselves as monospaced when
   in fact they are dual-spaced (half and full).  This makes
   FreeType very confused as it forces all widths to match.
   Undo this magic by disabling the width forcing code -->
+
+<!-- Removing as this is irrelevant for the stock device
 	<match target="font">
 		<test name="family"><string>GulimChe</string></test>
 		<edit name="globaladvance"><bool>false</bool></edit>
@@ -231,7 +363,7 @@
 		<test name="family"><string>GungsuhChe</string></test>
 		<edit name="globaladvance"><bool>false</bool></edit>
 	</match>
-
+-->
 <!-- 
 	The Bitstream Vera fonts have GASP entries suggesting that hinting be
 	disabled below 8 ppem, but FreeType ignores those, preferring to use
@@ -239,7 +371,7 @@
 	didn't include the right instructions in the 'prep' table. Fix this
 	by disabling hinting manually at smaller sizes (< 8ppem)
  -->
-
+ <!-- Removing as this is irrelevant for the stock device
 	<match target="font">
 		<test name="family">
 			<string>Bitstream Vera Sans</string>
@@ -276,6 +408,7 @@
 		</edit>
 	</match>
 
+-->
 <!--
   Load per-user customization file
 -->
@@ -293,6 +426,7 @@
 	<alias>
 		<family>serif</family>
 		<prefer>
+			<family>NewTimes</family>
 			<family>DejaVu Serif</family>
 			<family>Bitstream Vera Serif</family>
 			<family>Times New Roman</family>
@@ -313,6 +447,8 @@
 	<alias>
 		<family>sans-serif</family>
 		<prefer>
+			<family>SwissA</family>
+			<family>Nokia Sans</family>
 			<family>DejaVu Sans</family>
 			<family>Bitstream Vera Sans</family>
 			<family>Verdana</family>
@@ -335,6 +471,7 @@
 	<alias>
 		<family>monospace</family>
 		<prefer>
+			<family>NewCourier</family>
 			<family>DejaVu Sans Mono</family>
 			<family>Bitstream Vera Sans Mono</family>
 			<family>Andale Mono</family>
