diff -Naur fontconfig/conf.d/20-newtimes-nohint.conf fontconfig-2.4.1-newtimes-nohint/conf.d/20-newtimes-nohint.conf
--- fontconfig/conf.d/20-newtimes-nohint.conf	1970-01-01 02:00:00.000000000 +0200
+++ fontconfig-2.4.1-newtimes-nohint/conf.d/20-newtimes-nohint.conf	2006-09-28 15:44:41.000000000 +0300
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
+<!-- conf.d/sub-pixel.conf -->
+<fontconfig>
+<!-- 
+	Fix the Freetype hinting problem with NewCourier's "#" chars	
+ -->
+
+	<match target="font">
+		<test name="family">
+			<string>NewCourier</string>
+		</test>
+		<edit name="hinting">
+			<bool>false</bool>
+		</edit>
+	</match>
+
+</fontconfig>
diff -Naur fontconfig/conf.d/Makefile.am fontconfig-2.4.1-newtimes-nohint/conf.d/Makefile.am
--- fontconfig/conf.d/Makefile.am	2006-09-10 07:07:07.000000000 +0300
+++ fontconfig-2.4.1-newtimes-nohint/conf.d/Makefile.am	2006-09-28 15:46:45.000000000 +0300
@@ -33,6 +33,7 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
+	20-newtimes-nohint.conf \
 	30-amt-aliases.conf \
 	30-urw-aliases.conf \
 	40-generic.conf \
@@ -52,6 +53,7 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
+	20-newtimes-nohint.conf \
 	30-amt-aliases.conf \
 	30-urw-aliases.conf \
 	40-generic.conf \
diff -Naur fontconfig/conf.d/Makefile.in fontconfig-2.4.1-newtimes-nohint/conf.d/Makefile.in
--- fontconfig/conf.d/Makefile.in	2006-09-15 20:12:37.000000000 +0300
+++ fontconfig-2.4.1-newtimes-nohint/conf.d/Makefile.in	2006-09-28 15:47:02.000000000 +0300
@@ -242,6 +242,7 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
+	20-newtimes-nohint.conf \
 	30-amt-aliases.conf \
 	30-urw-aliases.conf \
 	40-generic.conf \
@@ -261,6 +262,7 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
+	20-newtimes-nohint.conf \
 	30-amt-aliases.conf \
 	30-urw-aliases.conf \
 	40-generic.conf \
