diff -Naur fontconfig/conf.d/40-generic.conf fontconfig-2.4.1-optimize-config/conf.d/40-generic.conf
--- fontconfig/conf.d/40-generic.conf	2006-09-10 06:33:43.000000000 +0300
+++ fontconfig-2.4.1-optimize-config/conf.d/40-generic.conf	2006-09-28 16:44:39.000000000 +0300
@@ -10,6 +10,7 @@
   Serif faces
  -->
 	<alias>
+	 	<family>NewTimes</family>
 		<family>Bitstream Vera Serif</family>
 		<family>DejaVu Serif</family>
 		<family>Times New Roman</family>
@@ -29,6 +30,8 @@
   Sans-serif faces
  -->
 	<alias>
+	 	<family>SwissA</family>
+		<family>Nokia Sans</family>
 		<family>Bitstream Vera Sans</family>
 		<family>DejaVu Sans</family>
 		<family>Helvetica</family>
@@ -50,7 +53,8 @@
 <!--
   Monospace faces
  -->
- 	<alias>
+	<alias>
+		<family>NewCourier</family> 
 		<family>Bitstream Vera Sans Mono</family>
 		<family>DejaVu Sans Mono</family>
 		<family>Courier</family>
--- fontconfig/conf.d/20-newtimes-nohint.conf   1970-01-01 02:00:00.000000000 +0200
+++ fontconfig-2.4.1-newtimes-nohint/conf.d/20-newtimes-nohint.conf     2006-09-28 16:38:32.000000000 +0300
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
+<!-- conf.d/sub-pixel.conf -->
+<fontconfig>
+<!--
+       Fix the Freetype hinting problem with NewCourier's "#" chars
+ -->
+
+       <match target="font">
+               <test name="family">
+                       <string>NewCourier</string>
+               </test>
+               <edit name="hinting">
+                       <bool>false</bool>
+               </edit>
+       </match>
+
+</fontconfig>
diff -Naur fontconfig/conf.d/60-latin.conf fontconfig-2.4.1-optimize-config/conf.d/60-latin.conf
--- fontconfig/conf.d/60-latin.conf	2006-09-10 06:30:00.000000000 +0300
+++ fontconfig-2.4.1-optimize-config/conf.d/60-latin.conf	2006-09-28 16:44:39.000000000 +0300
@@ -4,6 +4,7 @@
 	<alias>
 		<family>serif</family>
 		<prefer>
+			<family>NewTimes</family>
 			<family>Bitstream Vera Serif</family>
 			<family>DejaVu Serif</family>
 			<family>Times New Roman</family>
@@ -16,6 +17,8 @@
 	<alias>
 		<family>sans-serif</family>
 		<prefer>
+			<family>SwissA</family>
+			<family>Nokia Sans</family>
 			<family>Bitstream Vera Sans</family>
 			<family>DejaVu Sans</family>
 			<family>Verdana</family>
@@ -29,6 +32,7 @@
 	<alias>
 		<family>monospace</family>
 		<prefer>
+			<family>NewCourier</family>
 			<family>Bitstream Vera Sans Mono</family>
 			<family>DejaVu Sans Mono</family>
 			<family>Andale Mono</family>
diff -Naur fontconfig/conf.d/Makefile.am fontconfig-2.4.1-optimize-config/conf.d/Makefile.am
--- fontconfig/conf.d/Makefile.am	2006-09-10 07:07:07.000000000 +0300
+++ fontconfig-2.4.1-optimize-config/conf.d/Makefile.am	2006-09-28 16:46:34.000000000 +0300
@@ -33,6 +33,7 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
+	20-newtimes-nohint.conf \
 	30-amt-aliases.conf \
 	30-urw-aliases.conf \
 	40-generic.conf \
@@ -52,17 +53,14 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
-	30-amt-aliases.conf \
-	30-urw-aliases.conf \
+	20-newtimes-nohint.conf \
 	40-generic.conf \
 	49-sansserif.conf \
 	50-user.conf \
 	51-local.conf \
 	60-latin.conf \
-	65-fonts-persian.conf \
 	65-nonlatin.conf \
 	69-unifont.conf \
-	80-delicious.conf \
 	90-synthetic.conf
 
 EXTRA_DIST = $(CONF_FILES)
diff -Naur fontconfig/conf.d/Makefile.in fontconfig-2.4.1-optimize-config/conf.d/Makefile.in
--- fontconfig/conf.d/Makefile.in	2006-09-15 20:12:37.000000000 +0300
+++ fontconfig-2.4.1-optimize-config/conf.d/Makefile.in	2006-09-28 16:46:49.000000000 +0300
@@ -242,6 +242,7 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
+	20-newtimes-nohint.conf \
 	30-amt-aliases.conf \
 	30-urw-aliases.conf \
 	40-generic.conf \
@@ -261,17 +262,14 @@
 	20-fix-globaladvance.conf \
 	20-lohit-gujarati.conf \
 	20-unhint-small-vera.conf \
-	30-amt-aliases.conf \
-	30-urw-aliases.conf \
+	20-newtimes-nohint.conf \
 	40-generic.conf \
 	49-sansserif.conf \
 	50-user.conf \
 	51-local.conf \
 	60-latin.conf \
-	65-fonts-persian.conf \
 	65-nonlatin.conf \
 	69-unifont.conf \
-	80-delicious.conf \
 	90-synthetic.conf
 
 EXTRA_DIST = $(CONF_FILES)
