Index: gtk+/gtk/gtkcellrendererpixbuf.c
===================================================================
--- gtk+/gtk/gtkcellrendererpixbuf.c	(revision 956)
+++ gtk+/gtk/gtkcellrendererpixbuf.c	(working copy)
@@ -210,6 +210,10 @@
 
   if (cellpixbuf->pixbuf)
     g_object_unref (cellpixbuf->pixbuf);
+  if (cellpixbuf->pixbuf_expander_open)
+    g_object_unref (cellpixbuf->pixbuf_expander_open);
+  if (cellpixbuf->pixbuf_expander_closed)
+    g_object_unref (cellpixbuf->pixbuf_expander_closed);
 
   g_free (priv->stock_id);
   g_free (priv->stock_detail);
