Index: gtk+/gtk/gtkicontheme.c
===================================================================
--- gtk+/gtk/gtkicontheme.c	(revision 1245)
+++ gtk+/gtk/gtkicontheme.c	(working copy)
@@ -1824,19 +1831,6 @@
   min_dir = NULL;
   has_larger = FALSE;
 
-  /* Builtin icons are logically part of the default theme and
-   * are searched before other subdirectories of the default theme.
-   */
-  if (strcmp (theme->name, DEFAULT_THEME_NAME) == 0 && use_builtin)
-    {
-      closest_builtin = find_builtin_icon (icon_name, size,
-					   &min_difference,
-					   &has_larger);
-
-      if (min_difference == 0)
-	return icon_info_new_builtin (closest_builtin);
-    }
-
   l = theme->dirs;
   while (l != NULL)
     {
@@ -1859,7 +1853,6 @@
 		{
 		  min_difference = difference;
 		  min_dir = dir;
-		  closest_builtin = NULL;
 		  has_larger = smaller;
 		}
 	    }
@@ -1869,7 +1862,6 @@
 		{
 		  min_difference = difference;
 		  min_dir = dir;
-		  closest_builtin = NULL;
 		}
 	    }
 
@@ -1878,9 +1870,6 @@
       l = l->next;
     }
 
-  if (closest_builtin)
-    return icon_info_new_builtin (closest_builtin);
-  
   if (min_dir)
     {
       GtkIconInfo *icon_info = icon_info_new ();
@@ -1929,6 +1918,19 @@
       return icon_info;
     }
  
+  /* When an icon isn't found even in the default theme, try builtin stock
+   * icons as the last resort
+   */
+  if (strcmp (theme->name, DEFAULT_THEME_NAME) == 0 && use_builtin)
+    {
+      closest_builtin = find_builtin_icon (icon_name, size,
+					   &min_difference,
+					   &has_larger);
+
+      if (closest_builtin)
+	return icon_info_new_builtin (closest_builtin);
+    }
+  
   return NULL;
 }
 
