--- gtk+/gtk/gtktreeview.c	(revision 1324)
+++ gtk+/gtk/gtktreeview.c	(revision 1325)
@@ -3001,9 +3001,12 @@
 
   if (gtk_tree_row_reference_valid (tree_view->priv->queued_select_row))
     {
-      gtk_tree_view_real_set_cursor (tree_view,
-                                     gtk_tree_row_reference_get_path (tree_view->priv->queued_select_row),
-                                     TRUE, TRUE);
+      GtkTreePath *select_path =
+        gtk_tree_row_reference_get_path (tree_view->priv->queued_select_row);
+
+      gtk_tree_view_real_set_cursor (tree_view, select_path, TRUE, TRUE);
+
+      gtk_tree_path_free (select_path);
       gtk_tree_row_reference_free (tree_view->priv->queued_select_row);
       tree_view->priv->queued_select_row = NULL;
     }
@@ -3860,9 +3863,9 @@
           gtk_tree_row_reference_free (tree_view->priv->last_drag_row);
           tree_view->priv->last_drag_row =
             gtk_tree_row_reference_new (tree_view->priv->model, path);
-          gtk_tree_path_free (path);
         }
 
+      gtk_tree_path_free (path);
       gtk_tree_path_free (last_drag_path);
     }
 
