--- gtk+-2.6.4/gtk/gtkmenushell.h	2004-12-11 00:09:22.000000000 +0200
+++ gtk+-2.6.4/gtk/gtkmenushell.h	2005-06-27 13:41:44.306377552 +0300
@@ -56,7 +56,7 @@
   GtkWidget *active_menu_item;
   GtkWidget *parent_menu_shell;
   
-  guint button;
+  guint button;                 /* unused */
   guint32 activate_time;
 
   guint active : 1;
@@ -124,6 +124,12 @@
 void _gtk_menu_shell_remove_mnemonic (GtkMenuShell *menu_shell,
 				      guint         keyval,
 				      GtkWidget    *target);
+gboolean gtk_menu_shell_get_take_focus (GtkMenuShell *menu_shell);
+void     gtk_menu_shell_set_take_focus (GtkMenuShell *menu_shell,
+                                        gboolean      take_focus);
+
+/* Private function */
+void _gtk_menu_shell_set_first_click (GtkMenuShell *menu_shell);
 
 #ifdef __cplusplus
 }
