--- gtk+-2.6.4/gtk/gtktextview.h	2004-12-08 23:25:08.000000000 +0200
+++ gtk+-2.6.4/gtk/gtktextview.h	2005-06-27 13:41:45.628176608 +0300
@@ -95,8 +95,7 @@
 
   guint accepts_tab : 1;
   
-  /* this flag is no longer used */
-  guint reserved : 1;
+  guint width_changed : 1;
   
   /* debug flag - means that we've validated onscreen since the
    * last "invalidate" signal from the layout
@@ -104,7 +103,7 @@
   guint onscreen_validated : 1;
 
   guint mouse_cursor_obscured : 1;
-  
+
   GtkTextWindow *text_window;
   GtkTextWindow *left_window;
   GtkTextWindow *right_window;
