Index: modules/http-neon-method.c
===================================================================
--- modules/http-neon-method.c	(revision 375)
+++ modules/http-neon-method.c	(arbetskopia)
@@ -955,7 +955,8 @@ strip_semicolon (const char *value)
 static void
 set_content_type (GnomeVFSFileInfo *file_info, const char *value)
 {
-	g_free (file_info->mime_type);
+	if (file_info->mime_type)
+		return;
 
 	file_info->mime_type = strip_semicolon (value);
 	file_info->valid_fields |= GNOME_VFS_FILE_INFO_FIELDS_MIME_TYPE;
@@ -1840,9 +1841,17 @@ http_get_file_info (HttpContext *context
 	ne_propfind_handler *pfh;
 	ne_request *req;
 	int res;
+	const gchar *mime_type;
 
 	DEBUG_HTTP_CONTEXT (context);
 	
+	/* Let's fill in the mime type first */
+	mime_type = gnome_vfs_mime_type_from_name_or_default (context->path, NULL);
+	if (!mime_type || strcmp (mime_type, "application/octet-stream") != 0) {
+		info->mime_type = g_strdup (mime_type);
+		info->valid_fields |= GNOME_VFS_FILE_INFO_FIELDS_MIME_TYPE;
+	}
+
 	/* no dav server */
 	if (context->dav_mode == FALSE || context->dav_class == NO_DAV) 
 		goto head_start;
