? compile
? gnome-vfs-zip
? index.html
Index: modules/http-neon-method.c
===================================================================
RCS file: /cvs/gnome/gnome-vfs/modules/http-neon-method.c,v
retrieving revision 1.40
diff -u -p -r1.40 http-neon-method.c
--- modules/http-neon-method.c	15 Sep 2006 19:31:00 -0000	1.40
+++ modules/http-neon-method.c	10 Oct 2006 11:12:17 -0000
@@ -1104,9 +1104,7 @@ std_headers_to_file_info (ne_request *re
 	
 	value  = ne_get_response_header (req, "Content-Type");
 
-	if (value != NULL) {
-		g_free (info->mime_type);
-
+	if (value != NULL && info->mime_type == NULL) {
 		info->mime_type = strip_semicolon (value);
 		info->valid_fields |= GNOME_VFS_FILE_INFO_FIELDS_MIME_TYPE;
 	}
@@ -1827,9 +1825,17 @@ http_get_file_info (HttpContext *context
 	ne_propfind_handler *pfh;
 	ne_request *req;
 	int res;
+	const gchar *mime_type;
 
 	DEBUG_HTTP_CONTEXT (context);
 	
+	/* Let's fill in the mime type first */
+	mime_type = gnome_vfs_mime_type_from_name_or_default (context->path, NULL);
+	if (!mime_type || strcmp (mime_type, "application/octet-stream") != 0) {
+		info->mime_type = g_strdup (mime_type);
+		info->valid_fields |= GNOME_VFS_FILE_INFO_FIELDS_MIME_TYPE;
+	}
+
 	/* no dav server */
 	if (context->dav_mode == FALSE || context->dav_class == NO_DAV) 
 		goto head_start;
