--- libgnomevfs/gnome-vfs-volume-monitor-daemon.c.orig	2006-09-21 12:03:00.000000000 +0200
+++ libgnomevfs/gnome-vfs-volume-monitor-daemon.c	2006-09-21 12:03:07.000000000 +0200
@@ -834,39 +834,42 @@
 
 /* Nokia patch */
 #define MMC_LABEL_FILE "/tmp/.mmc-volume-label"
+#define MMC_LABEL_FILE2 "/tmp/.internal-mmc-volume-label"
 #define MMC_LABEL_LENGTH 11
-#define MMC_LABEL_UNDEFINED_NAME "mmc-undefined-name"
+#define MMC_LABEL_UNDEFINED_NAME          "mmc-undefined-name"
+#define MMC_LABEL_UNDEFINED_NAME_INTERNAL "mmc-undefined-name-internal"
 
 static char *
-get_mmc_name_from_label_file (void)
+get_mmc_name_from_label_file (const char *label_file,
+			      gboolean    internal)
 {
 	int     fd;
 	ssize_t size_read;
 	char    buf[MMC_LABEL_LENGTH + 1];
 	
-	if (!g_file_test (MMC_LABEL_FILE, G_FILE_TEST_EXISTS)) {
-		return g_strdup (MMC_LABEL_UNDEFINED_NAME);
+	if (!g_file_test (label_file, G_FILE_TEST_EXISTS)) {
+		goto unknown;
 	}
 
-	fd = open (MMC_LABEL_FILE, O_RDONLY);
+	fd = open (label_file, O_RDONLY);
 	if (fd < 0) {
-		return g_strdup (MMC_LABEL_UNDEFINED_NAME);
+		goto unknown;
 	}
 
 	size_read = read (fd, buf, MMC_LABEL_LENGTH);
 	close (fd);
 	
 	if (size_read > 0) {
-		gchar *tmp;
-
 		buf[size_read] = '\0';
-
-		tmp = g_strndup (buf, size_read);
-
-		return tmp;
+		return g_strndup (buf, size_read);
 	}
-	
-	return g_strdup (MMC_LABEL_UNDEFINED_NAME);
+
+unknown:
+	if (internal) {
+		return g_strdup (MMC_LABEL_UNDEFINED_NAME_INTERNAL);
+	} else {
+		return g_strdup (MMC_LABEL_UNDEFINED_NAME);
+	}		
 }
 /* End of Nokia patch */
 
@@ -881,7 +884,18 @@
 	if (mmc_mount && g_str_has_prefix (mount_path, mmc_mount)) {
 		gchar *mmc_name;
 		
-		mmc_name = get_mmc_name_from_label_file ();
+		mmc_name = get_mmc_name_from_label_file (MMC_LABEL_FILE, FALSE);
+
+		if (mmc_name) {
+			return mmc_name;
+		}
+	}
+ 
+	mmc_mount = g_getenv ("INTERNAL_MMC_MOUNTPOINT");
+	if (mmc_mount && g_str_has_prefix (mount_path, mmc_mount)) {
+		gchar *mmc_name;
+		
+		mmc_name = get_mmc_name_from_label_file (MMC_LABEL_FILE2, TRUE);
 
 		if (mmc_name) {
 			return mmc_name;
