Index: libgnomevfs/gnome-vfs-xfer.c
===================================================================
--- libgnomevfs/gnome-vfs-xfer.c	(revision 6696)
+++ libgnomevfs/gnome-vfs-xfer.c	(working copy)
@@ -2429,8 +2429,12 @@
 		 * file system, so we just forge ahead and hope for the best 
 		 */
 		target_dir_uri = gnome_vfs_uri_get_parent ((GnomeVFSURI *)target_uri_list->data);
-		result = gnome_vfs_get_volume_free_space (target_dir_uri, &free_bytes);
-
+		if (strcmp (gnome_vfs_uri_get_scheme (target_dir_uri), "obex") != 0) {
+			result = gnome_vfs_get_volume_free_space (target_dir_uri, &free_bytes);
+		} else {
+			result = GNOME_VFS_ERROR_NOT_SUPPORTED;
+		}
+		
 		if (result == GNOME_VFS_OK) {
 			if (!move && !link && progress->progress_info->bytes_total > free_bytes) {
 				result = GNOME_VFS_ERROR_NO_SPACE;
@@ -2505,7 +2509,11 @@
 					 * Make sure we have space for the copies.
 					 */
 					target_dir_uri = gnome_vfs_uri_get_parent ((GnomeVFSURI *)merge_target_uri_list->data);
-					result = gnome_vfs_get_volume_free_space (target_dir_uri, &free_bytes);
+					if (strcmp (gnome_vfs_uri_get_scheme (target_dir_uri), "obex") != 0) {
+						result = gnome_vfs_get_volume_free_space (target_dir_uri, &free_bytes);
+					} else {
+						result = GNOME_VFS_ERROR_NOT_SUPPORTED;
+					}
 					
 					if (result == GNOME_VFS_OK) {
 						if (progress->progress_info->bytes_total > free_bytes) {
