Index: libgnomevfs/gnome-vfs-hal-mounts.c
===================================================================
--- libgnomevfs/gnome-vfs-hal-mounts.c	(revision 11974)
+++ libgnomevfs/gnome-vfs-hal-mounts.c	(arbetskopia)
@@ -655,6 +655,8 @@ _hal_volume_policy_check (GnomeVFSVolume
 	/* needs to be a mountable filesystem OR contain crypto bits OR audio disc OR blank disc */
 	if (! ((libhal_volume_get_fsusage (hal_volume) == LIBHAL_VOLUME_USAGE_MOUNTABLE_FILESYSTEM) ||
 	       (libhal_volume_get_fsusage (hal_volume) == LIBHAL_VOLUME_USAGE_CRYPTO) ||
+	       /* added in maemo to allow drives for unformatted cards: */
+	       (libhal_volume_get_fsusage (hal_volume) == LIBHAL_VOLUME_USAGE_UNKNOWN) ||
 	       libhal_volume_disc_has_audio (hal_volume) ||
 	       libhal_volume_disc_is_blank (hal_volume)))
 		goto out;
@@ -825,7 +827,7 @@ _hal_add_drive_without_volumes (GnomeVFS
 
 	/* doesn't make sense for devices without removable storage */
 	if (!libhal_drive_uses_removable_media (hal_drive)) {
-		goto out;
+		/*goto out; Patched out from maemo to allow drives for unformatted cards */
 	}
 	
  	dbus_error_init (&error);
