Index: libgnomevfs/gnome-vfs-monitor.c
===================================================================
--- libgnomevfs/gnome-vfs-monitor.c	(revision 12626)
+++ libgnomevfs/gnome-vfs-monitor.c	(working copy)
@@ -452,7 +452,16 @@ gnome_vfs_monitor_callback (GnomeVFSMeth
 				send_uri_changes_now (monitor_handle, uri, now);
 				callback_data->send_at = now;
 			} else {
-				callback_data->send_at = last_data->send_at + CONSECUTIVE_CALLBACK_DELAY;
+				time_t when;
+
+				/* Make sure we don't send events in
+				 * the future, handle the case where
+				 * the system time is set back.
+				 */
+				when  = MIN(last_data->send_at, now);
+				when += CONSECUTIVE_CALLBACK_DELAY;
+
+				callback_data->send_at = when;
 			}
 		}
 		
