--- dbus/dbus/dbus-sysdeps.c	2005-02-11 01:31:27.000000000 +0200
+++ dbus.work/dbus/dbus-sysdeps.c	2005-05-13 15:03:01.602708216 +0300
@@ -3323,15 +3323,22 @@
                         dbus_gid_t     gid,
                         DBusError     *error)
 {
-  /* setgroups() only works if we are a privileged process,
-   * so we don't return error on failure; the only possible
-   * failure is that we don't have perms to do it.
-   * FIXME not sure this is right, maybe if setuid()
-   * is going to work then setgroups() should also work.
-   */
-  if (setgroups (0, NULL) < 0)
-    _dbus_warn ("Failed to drop supplementary groups: %s\n",
-                _dbus_strerror (errno));
+  struct passwd *pw = getpwuid(uid);
+  if (pw == NULL)
+    {
+      dbus_set_error (error, _dbus_error_from_errno (errno),
+                     "getpwuid() for UID '%d' failed: %s", uid,
+                     _dbus_strerror (errno));
+      return FALSE;
+    }
+
+  if (initgroups (pw->pw_name, pw->pw_gid) < 0)
+    {
+      dbus_set_error (error, _dbus_error_from_errno (errno),
+                      "initgroups() failed: %s",
+                      _dbus_strerror (errno));
+      return FALSE;
+    }
   
   /* Set GID first, or the setuid may remove our permission
    * to change the GID
