diff -ur dbus/dbus/dbus-internals.c dbus.new/dbus/dbus-internals.c
--- dbus/dbus/dbus-internals.c	2006-01-03 17:38:53.845920720 +0200
+++ dbus.new/dbus/dbus-internals.c	2006-01-03 17:56:21.289685064 +0200
@@ -162,16 +162,18 @@
  *
  * @param format printf-style format string.
  */
-void
+inline void
 _dbus_warn (const char *format,
             ...)
 {
+#if 0
   /* FIXME not portable enough? */
   va_list args;
 
   va_start (args, format);
   vfprintf (stderr, format, args);
   va_end (args);
+#endif
 }
 
 static dbus_bool_t verbose_initted = FALSE;
