--- glib/dbus-gmain.c-orig	2005-12-06 10:51:46.000000000 +0100
+++ glib/dbus-gmain.c	2005-12-09 15:26:50.000000000 +0100
@@ -374,6 +374,7 @@
   source = g_timeout_source_new (dbus_timeout_get_interval (timeout));
   g_source_set_callback (source, timeout_handler, timeout, NULL);
   g_source_attach (source, dbus_source->context);
+  g_source_unref (source);
   
   dbus_timeout_set_data (timeout, GUINT_TO_POINTER (g_source_get_id (source)),
       			 NULL);
