--- dbus/bus/connection.c	2005-09-26 14:01:18.000000000 +0200
+++ dbus.work/bus/connection.c	2005-09-26 14:00:03.000000000 +0200
@@ -198,13 +198,9 @@
     retry:
       
       dbus_error_init (&error);
-        
-      transaction = NULL;
-      while (transaction == NULL)
-        {
-          transaction = bus_transaction_new (d->connections->context);
-          _dbus_wait_for_memory ();
-        }
+
+      while ((transaction = bus_transaction_new (d->connections->context)) == NULL)
+        _dbus_wait_for_memory ();
         
       if (!bus_service_remove_owner (service, connection,
                                      transaction, &error))
