diff -Nru gtkhtml-3.9.1/src/gtkhtml.c gtkhtml-3.9.1.imcontext-leak-fix/src/gtkhtml.c
--- gtkhtml-3.9.1/src/gtkhtml.c	2006-11-20 10:58:55.000000000 +0200
+++ gtkhtml-3.9.1.imcontext-leak-fix/src/gtkhtml.c	2006-11-20 11:02:08.000000000 +0200
@@ -761,6 +761,11 @@
 			html->priv->resize_cursor = NULL;
 		}
 
+		if (html->priv->im_context) {
+			g_object_unref (html->priv->im_context);
+			html->priv->im_context = NULL;
+		}
+
 		g_free (html->priv->content_type);
 		g_free (html->priv->base_url);
 		g_free (html->priv);
