diff -ru gtkhtml-3.9.1/src/htmltextslave.c gtkhtml-3.9.1.mod/src/htmltextslave.c
--- gtkhtml-3.9.1/src/htmltextslave.c	2007-02-14 13:21:52.000000000 +0200
+++ gtkhtml-3.9.1.mod/src/htmltextslave.c	2007-02-14 13:23:29.000000000 +0200
@@ -605,15 +605,12 @@
 				tmp_gi = pango_glyph_item_split (&glyph_item->glyph_item, slave->owner->text, split_index);
 
 				/* free the beginning we don't need */
-				pango_item_free (tmp_gi->item);
-				pango_glyph_string_free (tmp_gi->glyphs);
-				g_free (tmp_gi);
-				
+				pango_glyph_item_free (tmp_gi) ;
 			}
 
 			if (offset + item->num_chars > end_offset) {
 				/* need to cut the ending of current glyph item */
-				PangoGlyphItem tmp_gi1, *tmp_gi2;
+				PangoGlyphItem tmp_gi1, *tmp_gi2 = NULL;
 				int split_index;
 
 				if (glyph_item->type == HTML_TEXT_SLAVE_GLYPH_ITEM_PARENTAL) {
@@ -626,12 +623,13 @@
 					- (slave->owner->text + tmp_gi1.item->offset);
 				tmp_gi2 = pango_glyph_item_split (&tmp_gi1, slave->owner->text, split_index);
 
-				glyph_item->glyph_item = *tmp_gi2;
+				glyph_item->glyph_item.item   = pango_item_copy (tmp_gi2->item);
+				glyph_item->glyph_item.glyphs = pango_glyph_string_copy (tmp_gi2->glyphs) ;
 
 				/* free the tmp1 content and tmp2 container, but not the content */
 				pango_item_free (tmp_gi1.item);
 				pango_glyph_string_free (tmp_gi1.glyphs);
-				g_free (tmp_gi2);
+				pango_glyph_item_free (tmp_gi2);
 
 				glyph_item->type = HTML_TEXT_SLAVE_GLYPH_ITEM_CREATED;
 				glyph_item->widths = NULL;
