Index: cmdline/apt-key
===================================================================
--- cmdline/apt-key	(revision 9895)
+++ cmdline/apt-key	(working copy)
@@ -59,7 +59,7 @@
 
 case "$command" in
     add)
-        $GPG --quiet --batch --import "$1"
+        $GPG --quiet --batch --ignore-time-conflict --import "$1"
         echo "OK"
         ;;
     del|rm|remove)
Index: debian/rules
===================================================================
--- debian/rules	(revision 9926)
+++ debian/rules	(working copy)
@@ -216,6 +216,8 @@
 	cp debian/bugscript debian/$@/usr/share/bug/apt/script
 
 #	cp share/debian-archive.gpg debian/$@/usr/share/$@

+	mkdir -p debian/$@/etc/apt/apt.conf.d
+	echo 'Acquire::gpgv::Options {"--ignore-time-conflict";};' \
+	     >debian/$@/etc/apt/apt.conf.d/50gpgv
 
 #	head -n 500 ChangeLog > debian/ChangeLog
