--- pango-orig/pango/break.c	2005-01-04 21:20:15.000000000 +0200
+++ pango-1.8.1/pango/break.c	2006-07-26 16:27:48.000000000 +0300
@@ -789,8 +789,6 @@
               switch (break_op)
                 {
                 case BREAK_PROHIBITED:
-                  /* can't break here */                  
-                  attrs[i].is_char_break = FALSE;
                   break;
 
                 case BREAK_IF_SPACES:
