--- pango-orig/pango/glyphstring.c	2007-01-23 01:41:43.000000000 +0200
+++ pango-1.16.4/pango/glyphstring.c	2007-02-27 15:10:54.000000000 +0200
@@ -42,6 +42,7 @@
   string->space = 0;
   string->glyphs = NULL;
   string->log_clusters = NULL;
+  string->has_cache = FALSE;
 
   return string;
 }
@@ -58,6 +59,8 @@
 {
   g_return_if_fail (new_len >= 0);
 
+  string->has_cache = FALSE;
+
   while (new_len > string->space)
     {
       if (string->space == 0)
@@ -262,8 +265,18 @@
 			    PangoRectangle   *ink_rect,
 			    PangoRectangle   *logical_rect)
 {
-  pango_glyph_string_extents_range (glyphs, 0, glyphs->num_glyphs,
-				    font, ink_rect, logical_rect);
+	if (!glyphs->has_cache)
+	{
+		pango_glyph_string_extents_range (glyphs, 0, glyphs->num_glyphs, font, &glyphs->ink_rect, &glyphs->logical_rect);
+		
+		glyphs->has_cache = TRUE;
+	}
+	
+	if (ink_rect)
+		*ink_rect = glyphs->ink_rect;
+
+	if (logical_rect)
+		*logical_rect = glyphs->logical_rect;
 }
 
 /**
diff -ruN pango-orig/pango/pango-glyph.h pango-1.16.0/pango/pango-glyph.h
--- pango-orig/pango/pango-glyph.h	2002-11-08 21:00:52.000000000 +0200
+++ pango-1.16.4/pango/pango-glyph.h	2005-03-24 17:15:30.012688332 +0200
@@ -77,6 +77,10 @@
 
   /*< private >*/
   gint space;
+
+  gboolean has_cache;
+  PangoRectangle ink_rect;
+  PangoRectangle logical_rect;
 };
 
 #define PANGO_TYPE_GLYPH_STRING (pango_glyph_string_get_type ())
