--- dbus-glib-0.72/doc/Makefile.am	2006-09-27 15:27:24.000000000 +0300
+++ dbus-glib-0.72.new/doc/Makefile.am	2006-11-14 17:47:24.820057376 +0200
@@ -1 +1,10 @@
-SUBDIRS=reference
\ No newline at end of file
+SUBDIRS=reference
+
+install-data-local:
+	if test -d api; then \
+          mkdir -p $(DESTDIR)$(datadir)/doc/dbus-glib/html; \
+          for F in `find api/html -type f`; do \
+            B=`basename $$F`; \
+            $(INSTALL_DATA) $$F $(DESTDIR)$(datadir)/doc/dbus-glib/html/$$B; \
+          done \
+        fi
