diff -pur dbus-glib-0.72/configure.ac dbus-glib-0.72.new/configure.ac
--- dbus-glib-0.72/configure.ac	2006-10-25 23:53:55.000000000 +0300
+++ dbus-glib-0.72.new/configure.ac	2007-02-08 11:30:58.990797032 +0200
@@ -58,6 +58,9 @@ AC_ARG_ENABLE(gcov, AS_HELP_STRING([--en
 AC_ARG_WITH(xml, AS_HELP_STRING([--with-xml=[libxml/expat]],[XML library to use]))
 AC_ARG_WITH(test-socket-dir, AS_HELP_STRING([--with-test-socket-dir=[dirname]],[Where to put sockets for make check]))
 
+AM_CONDITIONAL(ENABLE_GTK_DOC, test x$enable_gtk_doc = xyes)
+AM_CONDITIONAL(GTK_DOC_USE_LIBTOOL, test -n "$LIBTOOL")
+
 dnl DBUS_BUILD_TESTS controls unit tests built in to .c files 
 dnl and also some stuff in the test/ subdir
 AM_CONDITIONAL(DBUS_BUILD_TESTS, test x$enable_tests = xyes)
